### DISCUSSION TITLE
Hdd periodically write

### DISCUSSION THREAD

**Discussion by Daneltones8** [netalertx/NetAlertX#684] [General]
There is some way to reduce or remove the HDD write. Each minute application write fixed amount of data.
I pasted here the screen of disk write and my configurations that show which every minute it should does nothing.
If someone can help... Thanks
![Screenshot_2024-05-25-15-45-40-023_com android chrome](https://github.com/jokob-sk/NetAlertX/assets/158753873/6ab615d3-d09f-4b22-a76d-546dfbe31b98)
![Screenshot_2024-05-25-15-48-46-229_com android chrome](https://github.com/jokob-sk/NetAlertX/assets/158753873/3709b354-32f8-4b5d-8856-5cdf829ec615)

**Comment by ingoratsdorf** [comment]
Agree, same here. There's some pretty crazy disk write going on.
![image](https://github.com/user-attachments/assets/927e11e7-bc8b-4d82-ac1e-1d0de8d2ae83)

Running Version: 08:41:33 - Dev

1) Looks like every few minutes, it creates backups of configs in the docker mapped config folder. It's a ~150KB file that gets written every few minutes. That's the CSV backup schedule, so I switched that to 'on new device'.
2) Switched db maintenance from every 30 min to 4am once a day.

Both above did not make any difference whatsoever.

3) Increased SQLite3 cache: PRAGMA cache_size = 9000;

Did not make a difference

4) Made sure temp tables are in memory and not disk: PRAGMA temp_store = MEMORY;

Did not make any difference

5) Set logging to minimal

Did not make any difference

**Reply by jokob-sk** [reply]
Thanks for the detailed investigation and testing. 

I assume this is caused by the API update method. Not sure how to optimize this though:

https://github.com/jokob-sk/NetAlertX/blob/a0db36cb2d5ceddf870691f5297e1550df82ee53/server/api.py#L18

**Reply by jokob-sk** [reply]
hi @ingoratsdorf and @Daneltones8 ,

I just pushed a new dev image (ready in ~15min) to try to decrease io ops on the HD by buffering logs a bit. Not sure if that will make a difference.

Also, can you please let me know how you are observing the Disk IO on a container? I'd like to set it up on my end so I can try to debug this further. Thanks in advance!

**Reply by ingoratsdorf** [reply]
I use proxmox with InfluxDB. Proxmox writes stats about system and containers to InfluxDB every 10sec.

**Comment by Daneltones8** [comment]
Thank you! I will try it as soon as I will have a bit of time.
About displaying disk IO, I use QNAP nas which has this functionalities
integrated on container station.

Il ven 13 dic 2024, 22:49 jokob-sk ***@***.***> ha scritto:

**Comment by ingoratsdorf** [comment]
As per https://github.com/jokob-sk/NetAlertX/issues/914#issuecomment-2556246010, the latest dev from today seem to have fixed the issue to the largest part. Still lots of writing considering there is a cached SQLite3 working in the background. Dunno if it's that or something else. Are you writing any temp files?

Homeassistant writes 100x more data to MariaDB but that has a waaay lower disk write access, about 1/30.
![image](https://github.com/user-attachments/assets/c39d570b-27e8-400f-a2fc-140c3745cb77)

**Reply by jokob-sk** [reply]
Truth be told - HomeAssistant has a bigger dev community vs only one dev for the core backend in NAX(me lol) and NAX is build on an not so scalable architecture from the upstream original project. In order to reduce DB locks, which were already occurring on the original project, in NAX I'm exposing a JSON API layer, however these file scan grow quite quickly. These regular writes are I think when these files are updated.  

With the next version I'm introducing a GraphQL server/layer for Settings and Devices. Other endpoints will hopefully migrated in future. Next could be accessing the DB directly via the GraphQL layer as it still relies on those json endpoints. 

I also think the SqliteDB isn't great for parallel R/W operations. 

I am trying to balance features vs scalability7 vs stability on the project, and my personal life lol 

Happy of course for anyone more skilled to help reduce these writes on improve on the architecture.

**Reply by ingoratsdorf** [reply]
No blame, no worries. :-)
I know that game.

Not sure if this would be an option, but if the API JSON files do not need to be persistent, could they exist only in ram?

As in
`mount -t tmpfs tmpfs /app/api`

`https://www.man7.org/linux/man-pages/man5/tmpfs.5.html`
